draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1900 @@
+They are executed the same way inside the worker but defined using a different
+API.
+
+Concurrency Budget
+~~~~~~~~~~~~~~~~~~
+
+Each worker has a concurrency budget for each type of task it can run.
+
+The budget for each type of task is defined by a minimal
+and an optional maximal concurrency.
+
+The maximal concurrency budget can be dynamic or fixed.
+Dynamic maximal concurrency strategies may be used to determine the maximum
+concurrency based on the load factor of the server, available network bandwidth
+or any other requirement the user may have.
+
+.. note::
+	 If a user specifies a concurrency of more than 10 for :ref:`draft/celery-5-high-level-architecture:CPU Bound Tasks` a warning log message is emitted.
+
+   Too many threads can cause task execution to grind down to a halt.
+
+If there are more tasks in the :ref:`draft/celery-5-high-level-architecture:Internal Tasks Queue`
+than what is currently the allowed maximum task concurrency we increase the
+current maximum by that number of tasks.
+After this increase, there will be a configurable cooldown period during which
+the worker will execute the new tasks.
+After the cooldown period, if there are still more tasks in the :ref:`draft/celery-5-high-level-architecture:Internal Tasks Queue`
+than the current maximum capacity we increase the maximum concurrency exponentially
+by a configurable exponent multiplied by the number of budget increases.
+The result is rounded up.
+
+This process goes on until we either reach the maximum concurrency budget for
+that type of tasks or if the number of tasks in
+:ref:`draft/celery-5-high-level-architecture:Internal Tasks Queue` is lower than
+the current maximum concurrency.
+
+If the current number of tasks is lower than the current maximal concurrency
+we decrease it to the number of tasks that are currently executing.
+
+This algorithm can be replaced or customized by the user.
+
+Internal Tasks
+++++++++++++++
+
+Celery defines internal tasks to ensure it's operation and to provide
+support for it's features.
+
+The exact API for each task will be determined in another CEP.
+
+This list of internal tasks is not final.
+Other internal tasks may be defined in other CEPs.
+
+SystemD Notify
+~~~~~~~~~~~~~~
+
+This task reports the status of the worker to the SystemD service which is
+running it.
+
+It uses the `sd_notify`_ protocol to do so.
+
+Retry Failed Boot Step
+~~~~~~~~~~~~~~~~~~~~~~
+
+This task responds to a :term:`Command Message` which instructs the worker
+to retry an optional
+:ref:`Boot Step <draft/celery-5-high-level-architecture:Boot Steps>`
+which has failed during the worker's initialization procedure.
+
+The Boot Step's execution will be retried a configured amount of times
+before giving up.
+
+By default this task's
+:ref:`Circuit Breaker <draft/celery-5-high-level-architecture:Circuit Breaking>`
+is configured to never prevent or automatically fail the execution of this task.
+
+Boot Steps
```

Comment:
I am not sure if this is the correct place to mention this, but perhaps it should be nice to account for the case of allowing the worker to execute and consume a single message in a serverless platform, such as [AWS Lambda](https://aws.amazon.com/lambda/) and [Google Cloud Functions](https://cloud.google.com/functions/). Of course all other monitoring capabilities might not be available, but are instead provided by the platform. I think this would be a strong point / feature to be covered by Version 5.

What do you think?

Reply:
See Data Sources/Sinks.
I need to expand on the Router more than what's currently present.